### ISSUE TITLE
Parsing the target string "host-cuda" creates a JITModule and a cuda context, which is surprising

### ISSUE THREAD

**Issue by abadams** [halide/Halide#6448]
It does this because it needs a context (and a GPU device selected) to know what cuda capabilities the host system has in order to resolve it to something like x86-64-avx2-linux-cuda-cuda_capability_31

One way we could make this less surprising is delaying this resolution until the first time we actually need to know the cuda version, likely in lowering. We'd parse host-cuda into x86-64-avx2-linux-cuda-cuda_capability_host, and then at some point host would get converted into a concrete cuda capability. This is weird for different reasons though - you have this intermediate target. On the other hand, it would let you specify x86-64-avx2-linux-cuda-cuda_capability_host, which says sniff the cuda target but not the cpu. I don't think we can currently do that.

**Comment by abadams** [comment]
To state my case strongly: "host-cuda" without sniffing the host system's GPU is never the right target. It says "please assume I have a geforce gtx 580". It's very unlikely people are still running compute on a card released in 2010. I think if we really really don't want to sniff the host GPU by default, we'd have to make "host-cuda" an error: You must specify a compute capability.

**Comment by abadams** [comment]
Some data: The first time we use the cuda capability is to check if atomic adds are supported. This happens during lowering while offloading GPU loops. We also use it at that stage to set mcpu and mattrs

**Comment by zvookin** [comment]
I was going to mention sniffing the CUDA capability being necessary to support multitarget dispatch in ```halide_can_use_target_features``` but apparently we don't do that.

Basically all the ugly code in Target.cpp is there to make "host" appear magical in JIT context, which is perhaps actually used in our tests so...

My main thought on minimizing surprise is to either break out the fancy resolution support into a specific parsing call that does it, or if we don't want to break backward compatibility, offer a new parsing call that does not do that. One can then control where the CUDA sniffing happens or guarantee that a capability level is specified in the target, in which case sniffing for it won't happen.

Personally I'd be up for less magic internally here and a library routine and command line tool that does the fancy host resolution stuff. This will however break backward compatibility. (The command line tool would be in support of our tests.) In such a design, one would have the option of specifying the particular GPU in a system one wants to use.

**Comment by steven-johnson** [comment]
Yeah, breaking backwards compatibility is suboptimal but (I suspect) a tractable thing to deal with, since we now have versions and releases. What Zalman describes (API + cli tool) sounds appealing to me.

**Comment by steven-johnson** [comment]
> in the JIT context an API to resolve the host target is just boilerplate because you always use that Target.

This is definitely not the case; I'd argue that (at least in our tests) it's far more common to get the the target string from HL_JIT_TARGET.

**Comment by abadams** [comment]
Which is almost always "host" for real usage.

**Comment by zvookin** [comment]
You arguing that you want less surprise and then saying "an API to resolve the host target is just boilerplate because you always use that Target." The point of making it explicit is to give the caller the option to parse a target string without pulling in all the dependencies while still making it easy to get the full package if they want it. There are many ways to do this, but pushing the lazy evaluating the resolution without a way to make it happen early is not an improvement over what we have now.

**Comment by zvookin** [comment]
> The caller already has an option to parse a target string without pulling in all the dependencies. 

How?

**Comment by abadams** [comment]
Oh I think we were talking past each other. I was talking about parsing a target string without causing a JITSharedModule to be created.

**Comment by abadams** [comment]
This issue was opened because people complain that when they explicitly parse the Target string "host-cuda", a JITSharedModule gets created, and this is surprising. I'd rather not veer off into the merits of creating standalone host resolution utility, as it doesn't address the issue.

**Comment by abadams** [comment]
Approx summary of offline discussion (Zalman please correct me where I'm misremembering):

No change to default behavior.

We should add a method that *just* parses a Target string in the sense that it turns dash-separated tokens into a Target object with those flags set, and doesn't do any kind of filling-out or resolution on it. No turning on avx because "avx2" appears, for example. "host" would be an error for this method, because it doesn't correspond to any particular target enums. This method would be useful when the fully-elaborated target is known, and we just want to turn it from a string into a Target.

We should separately have methods that sniff the host target, and gets the host cuda capability. These already exist, I believe.

**Comment by zvookin** [comment]
That's basically it. I think "host" would for the features only part just fill in the CPU and OS, but otherwise not do much rather than being an error. It is offering two things the default method does as separate pieces of functionality. The idea being that this both helps surface what is going on by adding information to the documentation that is actually a part of the API, and anyone who doesn't want it (finds it "surprising") has an option to not get the features resolution support.
